File: addon/ast.py
"""
Ast of nodelang for use in blender (and prototype)
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
import typing
from typing import Any, Mapping, cast, Dict, List, Optional, ClassVar, Union, Sequence
from .bpy_wrap import bpy
import re
import unittest

from .parser import MaybeParsed, ParseContext, ParseError, ParseNonLexError, TokenizeErr
from . import token

# FIXME: in python3.11 add a primitive_types_raw list and unpack it into the Literal type below
primitive_types_raw = []

PrimitiveType = typing.Literal['f32', 'i32', 'u32', 'b8', 'bsdf']

# TODO: move types out of ast
primitive_types = ['f32', 'i32', 'u32']

@dataclass(slots=True)
class SerializeCtx:
  outstream: str = ""
  indent_level: int = 0
  # TODO: need to build the string here and keep track of how large the
  # current line is to control wrapping

  def write(self, s: str) -> None:
    """write out a string, (performing formatting implicitly in this subclass)"""
    pass

class Node(ABC):
  """
  A node in the Ast.

  NOTE: slots are not allowed since these need to be fully mutable, some upstream consumers
  [ab]use that precondition to dynamically restructure the ast while building it
  For performance might want to use slots and include a higher level referencing node that
  can have its type changed.
  """
  start: int
  end: int
  src: str

  def slice(self) -> str:
    return self.src[self.start:self.end]

  @abstractmethod
  def serialize(self, c: SerializeCtx = SerializeCtx()) -> str:
    pass

  @staticmethod
  def hardFinishParse(pctx: ParseContext, **ctx: Any) -> Union[ParseError, "Node"]:
    """
    Parse after consuming enough tokens to unambiguously expect the AST node type.
    The amount of already consumed tokens is dependent upon the AST node type.
    """
    raise NotImplementedError()

  def to_blender_node_args(self) -> Optional[Sequence[Mapping[str, Any]]]:
    raise TypeError(f'{type(self).__name__} does not coerce to a blender node')

@dataclass(unsafe_hash=True)
class Ident(Node):
  name: str
  quotes_not_needed_pattern: ClassVar[re.Pattern[str]] = re.compile(r'[a-zA-Z]\w*')

  def serialize(self, c: SerializeCtx = SerializeCtx()):
    # TODO: escape quotes and space and nonprintables
    quotes_not_needed = Ident.quotes_not_needed_pattern.fullmatch(self.name) is not None
    if quotes_not_needed: return self.name
    else: return f"'{self.name}'"
  
  @staticmethod
  def parse(pctx: ParseContext) -> MaybeParsed["Ident"]:
    tok = pctx.try_consume_tok_type(token.Type.ident)
    # TODO: create a zig-like _try function
    if tok is None or isinstance(tok, ParseError):
      return tok
    return Ident(cast(token.Ident, tok.tok).name)

class _TestIdent(unittest.TestCase):
  def test_parse(self):
    pctx = ParseContext("hello const")
    parsed = Ident.parse(pctx)
    self.assertIsNotNone(parsed)
    self.assertEqual("hello", parsed.name)

@dataclass
class _Named:
  name: Ident

class Named(_Named):
  def __init__(self, name: str | Ident):
    super().__init__(name if isinstance(name, Ident) else Ident(name))

@dataclass
class Struct(Named):
  """A compound datatype"""
  members: List[Union[PrimitiveType, "Struct"]] = field(default_factory=list)

Type = PrimitiveType | Struct

# including None for now since not yet sure how to represent an empty optional shader
PrimitiveValue = str | int | float | bool | None | List["PrimitiveValue"]

@dataclass
class Literal(Node):
  val: PrimitiveValue

  def serialize(self, c: SerializeCtx = SerializeCtx()) -> str:
    if isinstance(self.val, list):
      return f"[{', '.join(Literal.from_value(l).serialize(c) for l in self.val)}]"
    return str(self.val)

  @staticmethod
  def from_value(val: PrimitiveValue) -> "Literal":
    if not isinstance(val, (str, float, int, bool, list, type(None))):
      raise RuntimeError(f"unknown value '{val}' with type '{type(val)}' attempted to be used as a literal")
    return Literal(val)

  def to_blender_node_args(self):
    match self.val:
      case str(val):
        return { 'type': "ShaderNodeValue", value: self.val }
      case int(val) | float(val):
        return { 'type': "ShaderNodeValue", value: self.val }
      case bool(val):
        return { 'type': "ShaderNodeValue", value: int(self.val) }
      case None:
        return { 'type': "ShaderNodeValue", value: 0 }
      # FIXME: handle vec4, also probably use tuple?
      case [r, g, b, _a]:
        return { 'type': "ShaderNodeRGB", value: tuple(self.val) }
      case _:
        raise TypeError(f'Literal with value "{self.val}" had unhandled type when converting to node')

@dataclass
class VarRef(Node, Named):
  derefs: List[str] = field(default_factory=list) # maybe convert this to binary dot operators

  def serialize(self, c: SerializeCtx = SerializeCtx()):
    if not self.derefs:
      return self.name.serialize(c)
    else:
      return f'{self.name.serialize(c)}.{".".join(self.derefs)}'




class Expr(Node):
  """non-instantiable static method class"""
  # TODO: Expr = Literal | VarRef # | Call | BinOp
  # TODO: maybe don't allow None return?
  @staticmethod
  def parse(pctx: ParseContext) -> MaybeParsed["Expr"]:
    left = PrimaryExpr.parse(pctx)
    if left is None or isinstance(left, ParseError): return left
    return BinOp.hardFinishParse(pctx) or left


@dataclass
class NamedArg(Node, Named):
  val: Expr

  def serialize(self, c: SerializeCtx = SerializeCtx()):
    return f".{self.name.serialize(c)}={self.val.serialize(c)}"

@dataclass
class Call(Node, Named):
  args: List[NamedArg | Expr]

  def serialize(self, c: SerializeCtx = SerializeCtx()):
    arg_per_line = len(self.args) > 4
    nl = '\n'
    indent = '  ' # TODO: do real tree formatting
    return f'''{self.name.serialize(c)}({
      nl+indent if arg_per_line else ''
    }{(','+nl+indent if arg_per_line else ', ').join(a.serialize(c) for a in self.args)}{
      nl if arg_per_line else ''
    })'''


@dataclass
class BinOp(Node):
  Types = typing.Literal['+', '-', '*', '/', '^', '^^', '^/', '**', '&', '|', '&&', '||']

  tokens: ClassVar[Mapping[Types, token.Type]] = {
    '|': token.Type.pipe,
    '||': token.Type.pipePipe,
    '&&': token.Type.ampAmp,
    '&': token.Type.amp,
    '+': token.Type.plus,
    '-': token.Type.minus,
    '*': token.Type.star,
    '/': token.Type.fSlash,
    '^': token.Type.caret,
    '**': token.Type.starStar,
    '^^': token.Type.caretCaret,
    '^/': token.Type.caretFSlash,
  }

  _tokenList = tuple(tokens.values())

  precedences: ClassVar[Mapping[Types, int]] = {
    # '<' comparison will lower
    '|': 1,
    '||': 1,
    '&&': 2,
    '&': 2,
    '+': 3,
    '-': 4,
    '*': 5,
    '/': 6,
    '^': 7,
    '**': 8,
    '^^': 8,
    '^/': 9,
  }

  op: Types
  left: "Expr"
  right: "Expr"

  # TODO: instead of always wrapping in `()` that should be a part of the AST not of the serialization
  # TODO: print print using serialization context
  def serialize(self, c: SerializeCtx = SerializeCtx()):
    return f'({self.left.serialize(c)} {self.op} {self.right.serialize(c)})'

  # FIXME: this is not really a hardFinishParse since it can return just the left expr...
  @staticmethod
  def hardFinishParse(pctx: ParseContext, **ctx: Node) -> Union[ParseError, "Expr"]:
    """expects that the parser has already consumed the left expression"""
    left = ctx.get("left")
    if left is None: raise RuntimeError("BinOp.hardFinishParse called without a `left: Node` kwarg")

    cur_prec = 0
    while True:
      tok = pctx.consume_tok()
      if tok is None: return ParseNonLexError.UnexpectedEof
      if isinstance(tok, TokenizeErr): return tok
      # this looks wrong:
      if not token.Type.isinstance(tok, BinOp._tokenList): return ParseNonLexError.UnexpectedToken
      prec = BinOp.precedences[cast(BinOp.Types, tok.slice)]
      if prec < cur_prec:
        return cast(Expr, left)
      right = PrimaryExpr.parse(pctx)

      before_next_op = pctx.index
      maybe_next_op = pctx.consume_tok()
      if maybe_next_op is None: return ParseNonLexError.UnexpectedEof
      if isinstance(maybe_next_op, TokenizeErr): return maybe_next_op
      # this looks wrong:
      if not token.Type.isinstance(maybe_next_op, BinOp._tokenList): return ParseNonLexError.UnexpectedToken
      next_op = maybe_next_op
      next_op_prec = BinOp.precedences[cast(BinOp.Types, next_op.slice)]
      pctx.reset(before_next_op)

      if prec < next_op_prec:
        right = BinOp.hardFinishParse(pctx, left=right)

  def to_blender_node_args(self):
    return {
      'type': "ShaderNodeMath",
      'value': self.left,
      'value': self.right,
      operation: {
        '|': 'BITWISEOR',
        '||': 'OR',
        '&&': 'AND',
        '&': 'BITWISEAND',
        '+': 'ADD',
        '-': 'SUB',
        '*': 'MUL',
        '/': 'DIV',
        '^': 'BITWISEXOR',
        '**': 'EXP',
        '^^': 'POW',
        '^/': 'ROOT',
      }[self.op]
    }


@dataclass
class ConstDecl(Node):
  name: Ident
  value: Literal | VarRef | BinOp # TODO: this should really be an Expr sum type
  comment: Optional[str] = None
  type: Optional[Type] = None
  
  def serialize_type(self, c: SerializeCtx) -> str:
    if not self.type: return ''
    elif isinstance(self.type, Struct): return self.type.name.serialize(c)
    else: return self.type

  def serialize(self, c: SerializeCtx = SerializeCtx()):
    return (
      (f'/// {self.comment}' + '\n' if self.comment else '')
      + f'const {self.name.serialize(c)}'
      + (f': {self.serialize_type(c)}' if self.type else '')
      + f' = {self.value.serialize(c)}'
    )

  @staticmethod
  def parse(pctx: ParseContext) -> MaybeParsed["ConstDecl"]:
    start = pctx.index
    results: list[token.Token] = []

    # TODO: support Node types with parse methods, and parse any expression not just int
    for nodeOrTokType in (token.Type.const,
                          token.Type.ident,
                          token.Type.colon,
                          token.Type.ident,
                          token.Type.eq,
                          token.Type.int):
      parsed = pctx.try_consume_tok_type(nodeOrTokType)
      # TODO: create a zig-like _try function
      if parsed is None or isinstance(parsed, TokenizeErr):
        pctx.reset(start)
        return
      results.append(parsed)

    _, ident, _, _type_expr, _, val = results

    ident = Ident(cast(token.Ident, ident.tok).name)

    return ConstDecl(
      ident,
      Literal(cast(int, val)),
      None,
      # _type_expr # need to check the name against known types? (and parse full expressions)
    )

class _TestConstDecl(unittest.TestCase):
  def test_parse(self):
    pctx = ParseContext("const x: Test = 5")
    parsed = ConstDecl.parse(pctx)
    self.assertIsNotNone(parsed)
    self.assertEqual("x", parsed.name.name)

@dataclass
class StructAssignment(Node):
  variable: str
  field: Optional[str]
  value: Node


@dataclass
class Namespace(Node):
  # TODO: consider having it be a list of ConstDecl or Stmt
  decls: List[Node] = field(default_factory=list)
  decl_by_name: Dict[Ident, Node] = field(default_factory=dict)

  def append_decl(self, decl: ConstDecl) -> None:
    self.decls.append(decl)
    self.decl_by_name[decl.name] = decl
  
  def prepend_decl(self, new_decl: ConstDecl, target: Optional[ConstDecl] = None) -> None:
    index = 0 if target is None else self.decls.index(target)
    self.decls.insert(index, new_decl)
    self.decl_by_name[new_decl.name] = new_decl

  def serialize(self, c: SerializeCtx = SerializeCtx()):
    return '\n'.join(d.serialize(c) for d in self.decls)

  def to_blender_node_args(self):
      return [d.to_blender_node_args for d in self.decls]

# A group of declarationS
Group = Namespace

# The top level of the AST for a file
Module = Namespace

@dataclass
class ParenGroup(Node):
  inner: Node

  def serialize(self, c: SerializeCtx = SerializeCtx()) -> str:
    return f"({self.inner.serialize(c)})"

  @staticmethod
  def hardFinishParse(pctx: ParseContext) -> Union[ParseError, "ParenGroup"]:
    """assumes that an lPar `(` has already been parsed"""

    expr = Expr.parse(pctx)
    if expr is None: return ParseNonLexError.UnexpectedToken
    if isinstance(expr, ParseError): return expr

    rPar = pctx.try_consume_tok_type(token.Type.rPar)
    if isinstance(rPar, ParseError): return rPar
    if rPar is None: return ParseNonLexError.UnexpectedToken

    return ParenGroup(inner=expr)

@dataclass
class ArgExprList(Node):
  exprs: list[NamedArg | Expr]

  def serialize(self, c: SerializeCtx = SerializeCtx()) -> str:
    return ", ".join(e.serialize(c) for e in self.exprs)

  @staticmethod
  def hardFinishParse(pctx: ParseContext) -> Union[ParseError, "ArgExprList"]:
    """assumes the left parentheses has been parsed"""
    maybe_immediate_rpar = pctx.try_consume_tok_type(token.Type.rPar)
    if isinstance(maybe_immediate_rpar , TokenizeErr): return maybe_immediate_rpar
    if maybe_immediate_rpar is not None:
      return ArgExprList(exprs=[])

    exprs: list[NamedArg | Expr] = []
    while True:
      # TODO: lookahead for named arg here
      expr = Expr.parse(pctx)
      if expr is None: return ParseNonLexError.UnexpectedToken
      if isinstance(expr, ParseError): return expr
      if expr is None: return ParseNonLexError.UnexpectedEof
      exprs.append(expr)

      tok = pctx.try_consume_tok_type(token.Type.rPar, token.Type.comma)
      if isinstance(tok, TokenizeErr): return tok
      if tok is None: return ParseNonLexError.UnexpectedEof
      if token.Type.isinstance(tok, token.Type.rPar):
        break

    return ArgExprList(exprs)

# TODO: move the following to the parser module with more separation

class PrimaryExpr(Node): # TODO: should exclude binop
  @staticmethod
  def parse(pctx: ParseContext) -> MaybeParsed["PrimaryExpr"]:
    tok = pctx.consume_tok()
    if tok is None or isinstance(tok, TokenizeErr): return tok

    result = None
    match tok.tok:
      # looks like with a match expr I don't even really need Ident.parse
      case token.Ident(name):
        result = Ident(name)
        before_next = pctx.index
        next_tok = pctx.try_consume_tok_type(token.Type.lPar)
        if isinstance(next_tok, TokenizeErr): return next_tok
        elif next_tok is None:
          pctx.reset(before_next)
        else: # is lPar
          args = ArgExprList.hardFinishParse(pctx)
          if isinstance(args, ParseError):
            return args
          result = Call(result, args.exprs)
      case int(v) | float(v) | str(v) | bool(v):
        result = Literal(v)
      case token.Type.lPar:
        ParenGroup.hardFinishParse(pctx)
      case _:
        # would be better to raise an error here...
        return ParseNonLexError.UnexpectedToken

    return result

File: addon/ast.zig
//! parser for nodelang written in zig 

const std = @import("std");
const t = std.testing;
const debug = std.debug;

pub const Tok = union (enum) {
  ident,
  integer: i64,
  float: f64,
  string,
  lBrack, //[
  rBrack, //]
  lPar, //(
  rPar, //)
  @"const",
  dot,
  eq,
  colon,
  plus,
  minus,
  star,
  fSlash,
  caret,
  caretCaret,
  caretFSlash,
  starStar,
  amp, // &
  pipe,
  ampAmp, // &&
  pipePipe,
};

const Token = struct {
  tok: Tok,
  str: []const u8,

  pub fn new(tok: Tok, str: [] const u8) Token {
    return Token{.tok=tok, .str=str};
  }
};

pub const TokenizeErr = error {
  UnknownTok,
};

pub const ParseError = error {
  UnknownTok
};

pub const ParseContext = struct {
  source: []const u8,
  index: u64,

  fn new(source: []const u8) ParseContext { return ParseContext{.source=source, .index=0}; }

  fn slice(self: *ParseContext, start: usize, end: usize) []const u8 { return self.source[start..end]; }

  fn remaining_src(self: ParseContext) []const u8 { return self.source[self.index..]; }

  /// indexing but with optionals
  fn nth(self: ParseContext, n: usize) ?u8 {
    return if (n < self.remaining_src().len) self.remaining_src()[n] else null;
  }

  // TODO: allow starting and ending single quotes with escapes
  /// try to get the next token as if it's an identifier, assume unknown token if we fail
  /// - assumes whitespace has been skipped
  fn try_next_tok_keyword_or_ident(self: *ParseContext) TokenizeErr!Token {
    var i: usize = 1; // skip first char since it is assumed to be an identifier start
    var src: []const u8 = "";
    while (self.nth(i)) |c| {
      switch (c) {
        'a'...'z', 'A'...'Z', '0'...'9', '_' => { i += 1; },
        else => { src = self.remaining_src()[0..i]; break; }
      }
    }
    if (std.mem.eql(u8, src, "const")) {
      return Token.new(.@"const", self.remaining_src()[0..i]);
    } else {
      return Token.new(.ident, self.remaining_src()[0..i]);
    }
  }

  /// try to get the next token as if it's a number, assume unknown token if we fail
  /// - assumes whitespace has been skipped
  /// - assumes context starts with a digit
  fn try_next_tok_number(self: *ParseContext) TokenizeErr!?Token {
      // TODO: roll my own parser to not have redundant logic
      const hasPrefixChar = std.ascii.isAlpha(self.remaining_src()[1]) and std.ascii.isDigit(self.remaining_src()[2]);
      var hadPoint = false;
      var tokEnd: usize = 0;
      for (self.remaining_src()) |c, i| {
        if (c == '.') {
          hadPoint = true;
          continue;
        }
        if (c == '_') continue;
        const isPrefixChar = i == 1 and hasPrefixChar;
        if (!std.ascii.isDigit(c) and !isPrefixChar) {
          tokEnd = i;
          break;
        }
      }
      const src = self.remaining_src()[0..tokEnd];
      if (hadPoint) {
        const val = std.fmt.parseFloat(f64, src) catch return TokenizeErr.UnknownTok;
        return Token.new(.{.float=val}, src);
      } else {
        const val = std.fmt.parseInt(i64, src, 0) catch return TokenizeErr.UnknownTok;
        return Token.new(.{.integer=val}, src);
      }
  }

  // TODO: return ?enum{.eof}
  /// returns false if hit Eof
  fn skipAvailable(self: *ParseContext) bool {
    return while (self.nth(0)) |c| {
      switch (c) {
        ' ' , '\t' , '\n' => { self.index += 1; },
        else => break true,
      }
    } else false;
  }

  fn putBackSkips(self: *ParseContext) void {
    while (self.nth(0)) |c| {
      switch (c) {
        ' ' , '\t' , '\n' => { self.index -= 1; },
        else => return,
      }
    }
    return TokenizeErr.Eof;
  }

  fn consume_tok(self: *ParseContext) TokenizeErr!?Token {
    if (!self.skipAvailable()) return null;
    const maybeToken: TokenizeErr!?Token =
      if (std.mem.startsWith(u8, "^", self.remaining_src())) (
        if (std.mem.startsWith(u8, "^^", self.remaining_src())) Token.new(Tok.caretCaret, self.remaining_src()[0..2])
        else if (std.mem.startsWith(u8, "^/", self.remaining_src())) Token.new(Tok.caretFSlash, self.remaining_src()[0..2])
        else Token.new(Tok.caret, self.remaining_src()[0..1])
      )
      else if (std.mem.startsWith(u8, "+", self.remaining_src())) return Token.new(Tok.plus, self.remaining_src()[0..1])
      else if (std.mem.startsWith(u8, ":", self.remaining_src())) return Token.new(Tok.colon, self.remaining_src()[0..1])
      else if (std.mem.startsWith(u8, "-", self.remaining_src())) return Token.new(Tok.minus, self.remaining_src()[0..1])
      else if (switch (self.remaining_src()[0]) { 'a'...'z', 'A'...'Z', '_' => true, else => false })
        try self.try_next_tok_keyword_or_ident()
      else if (switch (self.remaining_src()[0]) { '0'...'9' => true, else => false })
        try self.try_next_tok_number()
      else null;
    const token = (try maybeToken) orelse return null;
    self.index += token.str.len;
    return token;
  }

  /// consume a token, if it is not of the given tag, put it back
  fn try_consume_tok_type(self: *ParseContext, tok_type: std.meta.Tag(Tok)) TokenizeErr!?Token {
    const start = self.index;
    const tok = (try self.consume_tok()) orelse return null;
    if (tok.tok == tok_type) {
      return tok;
    } else {
      self.reset(start);
      return null;
    }
  }

  fn reset(self: *ParseContext, index: usize) void {
    self.index = index;
  }
};

const Ident = struct {
  name: []const u8,
  // TODO: come up with a way to enforce the concept/comptime-interface of "Node"
  // maybe just a comptime function that ensures it, or one that adds it and generates a constructor for you
  srcSlice: []const u8,

  // FIXME: srcSlice is not always the name! once I added delimiter parsing, identifiers in nodelang can be `'` delimited which
  // is not in the name but is in the srcSlice
  pub fn new(name: []const u8) Ident { return Ident{.name=name }; } //.srcSlice=name };

  /// tries to parse an Ident out of the context
  fn parse(pctx: *ParseContext) ParseError!?Ident {
    const start = pctx.index;
    const tok = (try pctx.consume_tok()) orelse return null;
    if (tok.tok == Tok.ident) {
      return Ident{ .name = tok.str, .srcSlice = tok.str };
    } else {
      pctx.reset(start);
      return null;
    }
  }
};

test "parse Ident" {
  var pctx = ParseContext.new("hello const");
  const parsed = try Ident.parse(&pctx);
  try t.expect(parsed != null);
  try t.expectEqualStrings("hello", parsed.?.name);
}

fn TokOrNodeTypesToTuple(comptime tokOrNodeTypes: anytype) type {
  var fields: [tokOrNodeTypes.len]std.builtin.TypeInfo.StructField = undefined;
  for (tokOrNodeTypes) |tokOrNodeType, i| {
    const isParseable = @TypeOf(tokOrNodeType) == type;
    //@compileLog("tokOrNodeType = ", tokOrNodeType);
    //const isToken = @TypeOf(tokOrNodeType) == @typeInfo(Tok).Union.tag_type.?;
    fields[i] = std.builtin.TypeInfo.StructField{
      .name = std.fmt.comptimePrint("{}", .{i}),
      .field_type = if (isParseable) tokOrNodeType else Token,
      .default_value = null,
      .is_comptime = false,
      .alignment = 8,
    };
  }

  return @Type(std.builtin.TypeInfo{
    .Struct = .{
      .layout = std.builtin.TypeInfo.ContainerLayout.Auto,
      .fields = &fields,
      .decls = &.{},
      .is_tuple = true,
    }
  });
}

// FIXME: doesn't work due to a compiler error
/// parse several tokens, putting them all back if failing
fn parseMany(pctx: *ParseContext, comptime tokOrNodeTypes: anytype) TokenizeErr!?TokOrNodeTypesToTuple(tokOrNodeTypes) {
  var result: ?TokOrNodeTypesToTuple(tokOrNodeTypes) = undefined;
  inline for (tokOrNodeTypes) |tokOrNodeType, i| {
    const isParseable = @TypeOf(tokOrNodeType) == type;
    const isToken = @TypeOf(tokOrNodeType) == @typeInfo(Tok).Union.tag_type.?;
    var put_back_index_cuz_err: ?usize = null;

    const maybe_consumed =
      if (isParseable)
        tokOrNodeType.parse(pctx)
      else if (isToken) _: {
        // FIXME: eof is treated as an error here... which is wrong, see notes on TokenizeErr.Eof above
        const tok = try pctx.consume_tok();
        break :_ if (tok.tok == tokOrNodeType) tok else null;
      } else {
        @compileLog("bad tokOrNodeType = ", @TypeOf(tokOrNodeType));
        @compileError("parseMany list included something that was neither a token nor a parseable struct");
      };

    if (maybe_consumed) |consumed| {
      result.?[i] = consumed;
    } else  {
      // this could probably be done more elegantly by adding a local function and returning a real error here
      put_back_index_cuz_err = i;
      break;
    }

    if (put_back_index_cuz_err) |put_back_index| {
      // there was an error, clean up and return early
      var j: isize = @intCast(isize, put_back_index);
      while (j > 0) : (j -= 1) {
        // FIXME
        //pctx.put_back(result[j]);
      }

      result = null;
      break;
    }
  }
  return result;
}

const Decl = struct {
  ident: Ident,
  srcSlice: []const u8,

  fn parse(_pctx: *ParseContext) ?Decl {
    //if (parseMany(pctx, .{Tok.@"const", Ident, Tok.colon}) catch null) |toks| {

    const result = (struct {
      fn impl(pctx: *ParseContext) !?Decl {
        const srcStart = pctx.index;
        errdefer pctx.reset(srcStart);
        _  = (try pctx.try_consume_tok_type(.@"const")) orelse return error.PutBack;
        const ident = (try Ident.parse(pctx)) orelse return error.PutBack;
        _ = (try pctx.try_consume_tok_type(.colon)) orelse return error.PutBack;
        _ = (try pctx.try_consume_tok_type(.integer)) orelse return error.PutBack;

        const srcEnd = pctx.index;
        return Decl{ .ident=ident, .srcSlice=pctx.slice(srcStart, srcEnd) };
      }
    }).impl(_pctx);

    return result catch null;
  }
};

test "parse Decl" {
  var pctx = ParseContext.new("const x: Test = 5");
  const parsed = Decl.parse(&pctx);
  try t.expect(parsed != null);
  try t.expectEqualStrings("x", parsed.?.ident.name);
}

pub const BinOp = union (enum) {
  add,
  sub,
  dot
};

pub const Node = union (enum) {
  integer: i64,
  float: f64,
  @"bool": bool,
  string: []const u8,
  array: []const Node,
  call: struct { func: []const u8, args: []const Node },
  varRef: []const u8,
  binOp: struct { op: BinOp, left: *const Node, right: *const Node },
};
